src/v/cluster/archival/archiver_operations_impl.cc

```diff
@@ -877,28 +884,358 @@ class archiver_operations_impl : public archiver_operations_api {
     ss::abort_source _as;
     retry_chain_node _rtc;
     retry_chain_logger _rtclog;
-    ss::shared_ptr<cloud_storage_remote_api> _api;
-    ss::shared_ptr<segment_upload_builder_api> _upl_builder;
-    ss::shared_ptr<cluster_partition_manager_api> _pm;
+    std::unique_ptr<cloud_storage_remote_api> _api;
+    std::unique_ptr<segment_upload_builder_api> _upl_builder;
+    std::unique_ptr<cluster_partition_manager_api> _pm;
     config::binding<size_t> _read_buffer_size;
     config::binding<int16_t> _readahead;
     cloud_storage_clients::bucket_name _bucket;
     ss::scheduling_group _sg;
 };
 
 ss::shared_ptr<archiver_operations_api> make_archiver_operations_api(
-  ss::shared_ptr<cloud_storage_remote_api> remote,
-  ss::shared_ptr<cluster_partition_manager_api> pm,
-  ss::shared_ptr<segment_upload_builder_api> upl,
+  std::unique_ptr<cloud_storage_remote_api> remote,
+  std::unique_ptr<cluster_partition_manager_api> pm,
+  std::unique_ptr<segment_upload_builder_api> upl,
   cloud_storage_clients::bucket_name bucket) {
     return ss::make_shared<archiver_operations_impl>(
       std::move(remote),
       std::move(pm),
       std::move(upl),
-      ss::default_scheduling_group(), // TODO: use proper scheduling group
+      ss::default_scheduling_group(),
       std::move(bucket));
 }
 
 } // namespace detail
 
+struct one_time_stream_provider : public stream_provider {
+    explicit one_time_stream_provider(ss::input_stream<char> s)
+      : _st(std::move(s)) {}
+    ss::input_stream<char> take_stream() override {
+        auto tmp = std::exchange(_st, std::nullopt);
+        return std::move(tmp.value());
```

Comment:
I know you're likely using this correctly today, but misuse of this in the future is going to result in `bad_optional_access`. Better to add a safeguard or local error handling here now?